Feature/initial project setup with eslint, jest, prettier

nirajunify opened the pull request:
This PR contains commit which includes the following:
- Eslint setup
- Jest setup with dummy test in niraj directory
- Prettier setup for code formatting
Apart from these, I have also configured husky to ensure pre-commit and pre-push validation, which essentially means, your tests has to pass before you can commit or push code

- [x] Unit Test provided - Yes
- [x] Deprecated code removed? - Yes
- [x] Is Syntax & Formatting correct? - Yes
- [x] Is the approach to the problem appropriate? - Yes. Linting is required, and so is husky to ensure precommit validations
- [x] Can anything be simplified? - No
- [x] Is the code too specific to the product and needs generalization? - No. It is general enough
- [ ] Do forms need to sanitized? - NA
- [ ] Any obvious security flaws or potential holes? - NA
- [ ] Are the naming conventions appropriate? - No. Dummy test should have been named properly. It should be removed, I kept it for husky setup validation.
- [ ] Are there enough comments inline with the code? No
- [x] Is there documentation needed? - No, Guideline documents are enough.
- [x] Is there anything included that should not be? - No
- [x] Are all dependencies declared? - Yes
- [x] Are there any code smells? - No



Let me know if you have any question

Commit:
Project Setup/Tooling configuration commit
Installed and configured, typescript, prettier, eslint and configured all these tools.
Other than the above steps, I have also installed and configured jest for testing.
There is a sample js file and its relevant spec file in the niraj directory to test the jest configuration

Commit:
Husky installed and configured

Husky is a tool to allow us to use git hooks like pre-commit or pre-push.
Using these git hooks, we can validate whether the commit or push follow
best practices or not.
For example:
In this project, you cannot commit your code unless you fix all linting errors
Another restrictions are, you cannot push your code unless your tests passes

Commit:
Husky configured to use git hooks
Husky was configured and tested to use git hooks for pre-commit and pre-push validation.
Husky configuration will ensure that code gets committed only when it follows the guidelines and contains passing tests

praveen-unify commented on package.json:
```diff
@@ -0,0 +1,44 @@
+  "version": "1.0.0",
+  "description": "Coding Competition with TDD and TypeScript. \"Code Compete\" team will use this repository for practicing best practices while solving algorithmic challenges. Every day, a new challenge will be introduced to the team, team has to solve the challenge and commit their solution to this repository by making a PR to the master branch.",
+  "main": "jest.config.js",
+  "dependencies": {
```
The dependencies jest, ts-jest and typescript can be moved in devDependencies.

nirajunify replied:
Thank you Pareen for this good suggestion. I will implement this now.

krajasekhar-unify reviewed (dismissed):
Working well.

Commit:
Incorporated Praveen's recommendation about dev-dependencies
Praveen recommended to move eslint, and typescript to dev-dependecies section. I moved those items to their appropriate places.

krajasekhar-unify commented on package.json:
```diff
@@ -0,0 +1,43 @@
+  "husky": {
+    "hooks": {
+      "pre-commit": "npm run lint && npm test",
+      "pre-push": "npm run test"
```
Can we put this instead here?
```
      "pre-commit": "yarn lint:fix && yarn test",
      "pre-push": "yarn test"
```

nirajunify replied:
Yes we can. However, I would want this to done in separate issue. Will create and assign the issue to you.

krajasekhar-unify replied:
ok. I will raise pr there.

nirajunify replied:
No. I have created and assigned you a ticket. You should apply the changes in that ticket.

sudheer-sangaraju commented on package.json:
```diff
@@ -0,0 +1,43 @@
+  },
+  "husky": {
+    "hooks": {
+      "pre-commit": "npm run lint && npm test",
+      "pre-push": "npm run test"
```
@nirajunify Not sure whether we're trying to lint all the files. If we are running on all files, and if it is a big project it will take hell lot of time. Instead of that, it's better to use lint-staged which will run only on staged files instead of all files.

sudheer-sangaraju commented on src/niraj/foo.ts:
```diff
@@ -0,0 +1,2 @@
+export const sum = (...a: number[]): number =>
```
Can we provide more meaningful filename instead of foo.ts? ex: sum.ts etc
Same for variable names as well
